### ISSUE TITLE
[QUESTION] geography not valid

### ISSUE THREAD

**Issue by vforchi** [microsoft/mssql-jdbc#1031]
## Question
I believe I am doing something wrong, but when I use the Geography class to create WKB I get an invalid geography.

```
byte[] wkb = Geography.point(10, 20, 104001).serialize()
assert wkb[5] == 0x0c
```

This returns false, because that value is 0x08, meaning that the geometry is not valid, if I am not mistaken. But according to the CLR Types specification the valid bit should always be set for version 1.

**Comment by vforchi** [comment]
I have a similar problem with a polygon:
```
String wkt = "POLYGON((142.448916 -51.037471,142.41632 -52.52176,144.406309 -52.52273,144.37449 -51.038436,142.448916 -51.037471))"
byte[] wkb = Geometry.STGeomFromText(wkt, srid).serialize();
assert wkb[5] == 0x04;
```
Also, in this case wkb is 
```
41 96 01 00 01 00 05 00 00 00 d6 54 16 85 5d ce 61 40 32 c6 87 d9 cb 84 49 c0 77 15 52 7e 52 cd 61 40 34 2e 1c 8 c9 42 4a c0 41 62 bb 7b 0 d 62 40 ad 51 f d1 e8 42 4a c0 bb d5 73 d2 fb b 62 40 99 7e 89 78 eb 84 49 c0 d6 54 16 85 5d ce 61 40 32 c6 87 d9 cb 84 49 c0 01 00 00 00 02 00 00 00 00 01 00 00 00 ff ff ff ff 00 00 00 00 03
```
but I got the WKT from an existing geography in the DB, that has this representation:
```
41 96 01 00 01 04 05 00 00 00 32 c6 87 d9 cb 84 49 c0 d6 54 16 85 5d ce 61 40 34 2e 1c 08 c9 42 4a c0 77 15 52 7e 52 cd 61 40 ad 51 0f d1 e8 42 4a c0 41 62 bb 7b 00 0d 62 40 99 7e 89 78 eb 84 49 c0 bb d5 73 d2 fb 0b 62 40 32 c6 87 d9 cb 84 49 c0 d6 54 16 85 5d ce 61 40 01 00 00 00 02 00 00 00 00 01 00 00 00 ff ff ff ff 00 00 00 00 03
```
The length is in both cases 112, the difference is in the valid bit and in the data representation.

**Comment by peterbae** [comment]
Hi @vforchi, thanks for contacting us. The driver doesn't set the valid bit when the spatial datatype object is created from the user's input, because the driver doesn't verify the validity of  the actual object (however, when we send the object to the server and we retrieve it, the server would've set the valid bit if it is valid). Basically we let the server decide if the object is actually valid or not.

As for the polygon, could you show me how you're inserting the data, and how you're retrieving/viewing it?

**Comment by vforchi** [comment]
But the server is not setting the valid bit, should I do something special to enable it?

The original data was inserted with `STGeomFromText`, if I remember correctly. Now I am trying to insert the same data in another table from java using WKB.

A possible problem is that I am using a library that is supposed to work with many differed databases, so it does not use `SQLServerPreparedStatement#setGeography`, but setObject. I tried do pass the WKT string to setObject, but it complains that the data would be truncated.

I know this is a bit vague, let me know how I can provide more info. I have the feeling that the problem with the valid flag should be solved first.

**Comment by peterbae** [comment]
The default value for the valid flag is being set to false right now, but it can easily be changed if users need it to be true by default. Could you let me know why the valid flag needs to be true? Also, if you could post your Java code, I could help you better with the polygon problem as well.

**Comment by vforchi** [comment]
For the point I use this code:
```
byte[] wkb = Geography.point(lat, lon, 104001).serialize();
wkb[5] = (byte) 12;
```
And I write that into a `PreparedStatement` with `setObject` 
If I do not set the valid flag, the geography goes in and I can query the table, but if I join with another table on a geography column, using for example STIntersects(), the query succeeds but when I try to retrieve the ResultSet I get the following error:
```
Caused by: com.microsoft.sqlserver.jdbc.SQLServerException: A .NET Framework error occurred during execution of user-defined routine or aggregate "geography": 
System.ArgumentException: 24144: This operation cannot be completed because the instance is not valid. Use MakeValid to convert the instance to a valid instance. Note that MakeValid may cause the points of a geometry instance to shift slightly.
System.ArgumentException: 
.
	at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:256)
	at com.microsoft.sqlserver.jdbc.SQLServerResultSet$FetchBuffer.nextRow(SQLServerResultSet.java:5427)
	at com.microsoft.sqlserver.jdbc.SQLServerResultSet.fetchBufferNext(SQLServerResultSet.java:1758)
	at com.microsoft.sqlserver.jdbc.SQLServerResultSet.next(SQLServerResultSet.java:1016)
```

For the polygon, I use the code from my second post, and I get the following error when I try to insert the data:
```
Caused by: java.sql.BatchUpdateException: A .NET Framework error occurred during execution of user-defined routine or aggregate "geography": 
System.FormatException: One of the identified items was in an invalid format.
System.FormatException: 
.
	at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.executeBatch(SQLServerPreparedStatement.java:2083)
```

**Comment by peterbae** [comment]
Hi @vforchi, I agree that it would be best for the driver to set the valid bit to true (even if the driver isn't sure if it's valid), so that users can use these objects better. I'll be making a PR soon that addresses this issue, and let you test with it as well.

**Comment by peterbae** [comment]
I've made a PR (https://github.com/Microsoft/mssql-jdbc/pull/1035) and attached the driver in the comments. Could you test the jar and let me know if anything else needs fixing?

**Comment by vforchi** [comment]
Thanks, this fixed the problem with the points, but I still have the issue with the polygon.

When I try to insert this:
```
POLYGON((142.448916 -51.037471,142.41632 -52.52176,144.406309 -52.52273,144.37449 -51.038436,142.448916 -51.037471))
```

Which gets translated to 
```
41 96 1 0 1 4 5 0 0 0 d6 54 16 85 5d ce 61 40 32 c6 87 d9 cb 84 49 c0 77 15 52 7e 52 cd 61 40 34 2e 1c 8 c9 42 4a c0 41 62 bb 7b 0 d 62 40 ad 51 f d1 e8 42 4a c0 bb d5 73 d2 fb b 62 40 99 7e 89 78 eb 84 49 c0 d6 54 16 85 5d ce 61 40 32 c6 87 d9 cb 84 49 c0 1 0 0 0 2 0 0 0 0 1 0 0 0 ff ff ff ff 0 0 0 0 3
```

I get the following error:
```
Caused by: java.sql.BatchUpdateException: A .NET Framework error occurred during execution of user-defined routine or aggregate "geography": 
System.FormatException: One of the identified items was in an invalid format.
System.FormatException: 
```

**Comment by peterbae** [comment]
Hi @vforchi, thanks for confirming. From your above code, I see:

```
String wkt = "POLYGON((142.448916 -51.037471,142.41632 -52.52176,144.406309 -52.52273,144.37449 -51.038436,142.448916 -51.037471))"
byte[] wkb = Geometry.STGeomFromText(wkt, srid).serialize();
```

And I assume you're putting this into a Geography column? If that's the case, you should use Geography.STGeomFromText instead. I haven't done any complex operations with it, but I was able to insert this Geography data into a Geography column. Let me know if this solves your problem.

**Comment by vforchi** [comment]
Thanks, that was it. I completely overlooked it.

**Comment by peterbae** [comment]
Thanks for confirming. I'll keep this issue open for now, until PR #1035 is merged. I agree that it's better for the driver to send the valid bit set to true by default 👍

**Comment by peterbae** [comment]
Fixed in PR #1035. Closing issue.
